Optimize memory copy with bess::utils::Copy()

sangjinhan opened the pull request:
This PR introduces `bess::utils::Copy()`, which is a generic memory copy function. It's faster than `memcpy()` and DPDK's `rte_memcpy()`, so it is recommended to use this new function when performance is critical in the datapath code. This PR also replaces `memcpy()`/`rte_memcpy()` with the new function in such places.

The function works on both aligned/unaligned buffers. It provides an optional `sloppy` parameter, which generates slightly faster and _significantly smaller_ code when it is set. The sloppy version may copy additional up to 31 bytes to the end of the destination buffer to boost throughput. Since it differs from the original semantics of `memcpy()`, the sloppy option is disabled by default.

### Performance benchmark 
Measured with `core/utils/copy_bench.cc` on a machine with Xeon E5-2620 v4 @ 2.10GHz. The unit used in the table is Mops/s (thus higher is better). `dst` and `src` refer to the misalignment offset of each buffer from a CPU cacheline (64 bytes).

dst  | src  | size | Copy     | CopySloppy | rte_memcpy | memcpy
---  | ---  | ---- | -------- | -------    | ---------- | -------
0    | 0    | 4    |  333.63  | 333.633    |  285.97    | 166.802
0    | 0    | 7    |  310.397 |  333.63    | 250.223    | 166.802
0    | 0    | 8    |  333.632 | 333.632    | 285.971    | 181.964
0    | 0    | 14   |  306.643 | 333.633    | 250.203    | 181.965
46   | 0    | 14   |  288.048 | 234.535    | 250.224    | 181.961
50   | 0    | 14   |  302.559 | 234.515    | 250.223    | 181.965
0    | 0    | 18   |  333.634 | 333.633    | 333.632    | 153.969
46   | 0    | 18   |  333.635 | 234.645    | 333.632    | 153.969
50   | 0    | 18   |  192.081 | 234.603    | 166.337    | 153.971
0    | 10   | 31   |  333.632 |  333.63    | 333.632    | 153.971
0    | 0    | 32   |  333.632 | 333.632    | 333.631    |   166.8
15   | 19   | 48   |  323.82  | 331.596    |  285.97    | 166.802
0    | 0    | 48   |  333.632 | 326.689    | 249.901    | 166.802
2    | 0    | 60   |  333.632 | 147.066    | 250.156    | 166.802
0    | 0    | 64   |  333.632 | 328.049    | 249.896    | 166.797
0    | 14   | 64   |  316.969 | 332.494    | 250.223    | 166.643
0    | 18   | 64   |  319.015 | 332.578    | 250.224    | 166.781
0    | 0    | 100  |  196.842 |  217.97    | 134.763    | 143.515
0    | 0    | 128  |  144.775 | 218.417    | 111.021    | 83.5111
0    | 0    | 256  |  97.8924 | 132.083    |  76.984    | 92.0148
10   | 47   | 257  |  78.1008 | 89.6216    | 65.6321    | 69.9676
0    | 0    | 384  |  79.0841 | 95.6584    | 58.8706    | 80.5193
1    | 0    | 384  |  63.6995 | 66.5598    | 54.0979    | 69.5478
0    | 16   | 512  |  61.7649 | 75.4347    | 47.6571    | 71.4721
0    | 0    | 1024 |  58.2776 | 60.6538    | 27.0486    | 44.2116
0    | 14   | 1500 |  34.5575 | 34.6593    | 17.4057    | 28.5943
0    | 18   | 1500 |  34.6046 | 34.6435    | 17.1083    | 30.5496
0    | 0    | 1514 |  36.0912 | 40.8488    |   19.58    | 33.3589
0    | 0    | 1518 |  35.9363 | 40.8488    | 19.5822    | 33.3601
19   | 4    | 2047 |  24.5783 | 24.7114    | 13.5933    | 21.9958
0    | 0    | 4096 |  15.5164 | 15.5164    | 7.35315    |  11.705

nrdmao commented:
Some additional potentially interesting test cases:

Copy for L2 broadcast:
src=0 dst=0 size=1514, 1518

Copy for L3 multicast:
src=14, 18 dst=0 size=64, 1500

Copy for L2 encap:
src=0 dst=0, 46, 50 size=14, 18

sangjinhan commented:
Good idea. Added those data points to the table above.

apanda commented on core/utils/copy.h:
```diff
@@ -0,0 +1,204 @@
+
+  if (!sloppy && (bytes % block_size) != 0) {
+    size_t fringe = bytes % block_size;
+    dst_u = reinterpret_cast<decltype(dst_u)>(d + leftover_blocks);
```
Not a big deal but the use of `decltype(dst_u)` worries me here, since it is unambiguously `uintptr_t`. Is there a reason to use it here. I kind of understand for line 196 (it is much longer), but it seems strange here.

sangjinhan replied:
Yup, fixed.

sangjinhan commented:
I was originally concerned about the binary size, since the new function is always inlined. So I chose replace only `memcpy()` on data plane.

I added an non-inlined version as well, and updated the files you listed. Now there are two flavors:
* `CopyInlined()`: inlining is forced. For datapath use.
* `Copy()`: inlined only when appropriate. For generic use.

sangjinhan commented:
Rebased to master to resolve a conflict.

apanda commented:
Merged

![emrgencykittens_2017-apr-18](https://cloud.githubusercontent.com/assets/3251/25161512/36a32ad2-2471-11e7-8058-212203505c68.jpg)
